app/javascript/packages/memorable-date/index.ts

```diff
@@ -0,0 +1,386 @@
+    }
+    // Set another range error if applicable and messaging is available
+    return rangeErrors.find(
+      ({ min: rangeMin, max: rangeMax }) =>
+        (rangeMin && date < new Date(rangeMin)) || (rangeMax && date > new Date(rangeMax)),
+    )?.message;
+  }
+
+  /**
+   * @param attrName Name of attribute on this element containing a parseable date
+   * @returns Date Javascript object parsed from the given attribute or null if it's misssing/invalid
+   */
+  private getDateAttribute(attrName: string): Date | null {
+    const raw = this.getAttribute(attrName);
+    if (raw === null) {
+      return null;
+    }
+
+    const value = Date.parse(raw);
+    if (Number.isNaN(value)) {
+      return null;
+    }
+    return new Date(value);
+  }
+
+  /**
+   * Set a custom error message on the given fields for use by a containing
+   * ValidatedFieldElement.
+   *
+   * @param message Error message to display
+   * @param fields Fields to which the error message applies
+   */
+  private setValidity(message: string, ...fields: HTMLInputElement[]): void {
+    const { allInputs } = this;
+    allInputs.forEach((field) => {
+      if (fields.includes(field)) {
+        field.setCustomValidity(message);
+      } else {
+        field.setCustomValidity('');
+      }
+    });
+  }
+
+  /**
+   * @param errs Hash for looking up error messages
+   * @returns Function to show errors for missing values on a group of elements
+   */
+  private checkMissingValues(errs: MemorableDateErrorMessageLookup) {
+    return (fields: { [k: string]: HTMLInputElement | undefined }): boolean => {
+      const message = errs[`missing_${Object.keys(fields).join('_')}`];
+      const fieldValues = Object.values(fields);
+      if (fieldValues.every((field) => !field?.value)) {
+        message && this.setValidity(message, ...(fieldValues as HTMLInputElement[]));
+        return true;
+      }
+      return false;
+    };
+  }
+
+  /**
+   * @param errs Hash for looking up error messages
+   * @returns Function to show errors for invalid values on a group of elements
+   */
+  private checkFieldsInvalid(errs: MemorableDateErrorMessageLookup) {
+    return (fields: { [k: string]: HTMLInputElement | undefined }): boolean => {
+      const message = errs[`invalid_${Object.keys(fields).join('_')}`];
+      const fieldValues = Object.values(fields);
+      if (fieldValues.every((field) => field?.validity.patternMismatch)) {
+        message && this.setValidity(message, ...(fieldValues as HTMLInputElement[]));
+        return true;
+      }
+      return false;
+    };
+  }
+
+  /**
+   * @param entry Entry to check from Object.entries call
+   * @returns True if the given entry is a valid mapping for memorable date errors
+   */
+  private isValidErrorMessage(entry: [string, any]): entry is [MemorableDateErrorMessage, string] {
+    return (
+      Array.isArray(entry) &&
+      MemorableDateErrorMessage[entry[0]] === entry[0] &&
+      typeof entry[1] === 'string'
```

Comment:
Coming to this fresh, I wonder if it could help to give names to these key value pairs to improve readability for future maintainers.

Something like:

```ts
const [lookupKey, errorMessage] = entry;
return MemorableDateErrorMessage[lookupKey] === lookupKey && typeof errorMessage === 'string';
```

Reply:
Done for `errorMessage`.